File: core/src/main/cpp/keyx/KeyRequestData.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <io/MslEncoderException.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <keyx/KeyRequestData.h>
#include <io/MslObject.h>
#include <keyx/KeyExchangeFactory.h>
#include <MslEncodingException.h>
#include <MslError.h>
#include <MslKeyExchangeException.h>
#include <util/MslContext.h>
#include <util/MslUtils.h>
#include <string>

using namespace std;
using namespace netflix::msl;
using namespace netflix::msl::io;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace keyx {

namespace {
/** Key key exchange scheme. */
const string KEY_SCHEME = "scheme";
/** Key key request data. */
const string KEY_KEYDATA = "keydata";
} // namespace anonymous

//static
shared_ptr<KeyRequestData> KeyRequestData::create(shared_ptr<MslContext> ctx,
        shared_ptr<MslObject> keyRequestDataMo)
{
    try {
        // Pull the key data.
        const string schemeName = keyRequestDataMo->getString(KEY_SCHEME);
        const KeyExchangeScheme scheme = ctx->getKeyExchangeScheme(schemeName);
        if (scheme == KeyExchangeScheme::INVALID)
            throw MslKeyExchangeException(MslError::UNIDENTIFIED_KEYX_SCHEME, schemeName);
        shared_ptr<MslEncoderFactory> encoder = ctx->getMslEncoderFactory();
        shared_ptr<MslObject> keyData = keyRequestDataMo->getMslObject(KEY_KEYDATA, encoder);

        // Construct an instance of the concrete subclass.
        shared_ptr<KeyExchangeFactory> keyFactory = ctx->getKeyExchangeFactory(scheme);
        if (!keyFactory)
            throw MslKeyExchangeException(MslError::KEYX_FACTORY_NOT_FOUND, scheme.name());
        return keyFactory->createRequestData(ctx, keyData);
    } catch (const MslEncoderException& e) {
        throw MslEncodingException(MslError::MSL_PARSE_ERROR, "keyrequestdata " + keyRequestDataMo->toString(), e);
    }
}

shared_ptr<ByteArray> KeyRequestData::toMslEncoding(shared_ptr<MslEncoderFactory> encoder, const MslEncoderFormat& format) const
{
    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put(KEY_SCHEME, scheme.name());
    mo->put(KEY_KEYDATA, getKeydata(encoder, format));
    return encoder->encodeObject(mo, format);
}

bool KeyRequestData::equals(shared_ptr<const KeyRequestData> base) const
{
    if (!base) return false;
    if (this == base.get()) return true;
    return scheme == base->scheme;
}

bool operator==(const KeyRequestData& a, const KeyRequestData& b)
{
	shared_ptr<const KeyRequestData> ap(&a, &MslUtils::nullDeleter<KeyRequestData>);
	shared_ptr<const KeyRequestData> bp(&b, &MslUtils::nullDeleter<KeyRequestData>);
	return ap->equals(bp);
}

ostream& operator<<(ostream& os, const KeyRequestData& data)
{
	// FIXME
	return os << data.scheme.name();
}

ostream& operator<<(ostream& os, shared_ptr<KeyRequestData> data)
{
	return os << *data;
}

}}} // namespace netflix::msl::keyx

File: core/src/main/javascript/keyx/KeyRequestData.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>Key request data contains all the data needed to facilitate a exchange of
 * session keys with the requesting entity.</p>
 *
 * <p>Specific key exchange mechanisms should define their own key request data
 * types.</p>
 *
 * <p>Key request data is represented as
 * {@code
 * keyrequestdata = {
 *   "#mandatory" : [ "scheme", "keydata" ],
 *   "scheme" : "string",
 *   "keydata" : object
 * }} where:
 * <ul>
 * <li>{@code scheme} is the key exchange scheme</li>
 * <li>{@code keydata} is the scheme-specific key data</li>
 * </ul></p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
	"use strict";
	
	var MslEncodable = require('../io/MslEncodable.js');
	var AsyncExecutor = require('../util/AsyncExecutor.js');
	var MslKeyExchangeException = require('../MslKeyExchangeException.js');
	var MslEncoderException = require('../io/MslEncoderException.js');
	var MslEncodingException = require('../MslEncodingException.js');
	var MslError = require('../MslError.js');
	
    /**
     * Key key exchange scheme.
     * @const
     * @type {string}
     */
    var KEY_SCHEME = "scheme";
    /**
     * Key key request data.
     * @const
     * @type {string}
     */
    var KEY_KEYDATA = "keydata";

    var KeyRequestData = module.exports = MslEncodable.extend({
        /**
         * Create a new key request data object with the specified key exchange
         * scheme.
         *
         * @param {KeyExchangeScheme} scheme the key exchange scheme.
         */
        init: function init(scheme) {
            // The properties.
            var props = {
                keyExchangeScheme: { value: scheme, writable: false, configurable: false },
            };
            Object.defineProperties(this, props);
        },

        /**
         * @param {MslEncoderFactory} encoder MSL encoder factory.
         * @param {MslEncoderFormat} format MSL encoder format.
         * @param {{result: function(MslObject), error: function(Error)}}
         *        callback the callback that will receive the key data MSL
         *        representation or any thrown exceptions.
         * @throws MslEncoderException if there was an error constructing the MSL
         *         representation.
         */
        getKeydata: function(encoder, format, callback) {},
        
        /** @inheritDoc */
        toMslEncoding: function toMslEncoding(encoder, format, callback) {
            var self = this;
            
            this.getKeydata(encoder, format, {
                result: function(keydata) {
                    AsyncExecutor(callback, function() {
                        var mo = encoder.createObject();
                        mo.put(KEY_SCHEME, this.keyExchangeScheme.name);
                        mo.put(KEY_KEYDATA, keydata);
                        encoder.encodeObject(mo, format, callback);
                    }, self);
                },
                error: callback.error,
            }, this);
        },

        /**
         * @param {?} that the reference object with which to compare.
         * @return {boolean} true if the other object is a key request data
         *         with the same scheme.
         * @see #uniqueKey()
         */
        equals: function equals(that) {
            if (this === that) return true;
            if (!(that instanceof KeyRequestData)) return false;
            return this.keyExchangeScheme == that.keyExchangeScheme;
        },

        /**
         * @return {string} a string that uniquely identifies this key request
         *         data.
         * @see #equals(that)
         */
        uniqueKey: function uniqueKey() {
            return this.keyExchangeScheme;
        },
    });

    /**
     * Construct a new key request data instance of the correct type from the
     * provided MSL object.
     *
     * @param {MslContext} ctx MSL context.
     * @param {MslObject} keyRequestDataMo the MSL object.
     * @param {{result: function(KeyRequestData), error: function(Error)}}
     *        callback the callback will receive the key request data concrete
     *        instance or any thrown exceptions.
     * @throws MslEncodingException if there is an error parsing the data.
     * @throws MslCryptoException if there is an error verifying the key
     *         request data.
     * @throws MslEntityAuthException if the entity authentication data could
     *         not be created.
     * @throws MslKeyExchangeException if unable to create the key request
     *         data.
     */
    var KeyRequestData$parse = function KeyRequestData$parse(ctx, keyRequestDataMo, callback) {
        AsyncExecutor(callback, function() {
            try {
                // Pull the key data.
                var schemeName = keyRequestDataMo.getString(KEY_SCHEME);
                var scheme = ctx.getKeyExchangeScheme(schemeName);
                if (!scheme)
                    throw new MslKeyExchangeException(MslError.UNIDENTIFIED_KEYX_SCHEME, schemeName);
                var encoder = ctx.getMslEncoderFactory();
                var keyData = keyRequestDataMo.getMslObject(KEY_KEYDATA, encoder);

                // Construct an instance of the concrete subclass.
                var keyFactory = ctx.getKeyExchangeFactory(scheme);
                if (!keyFactory)
                    throw new MslKeyExchangeException(MslError.KEYX_FACTORY_NOT_FOUND, scheme.name);
                keyFactory.createRequestData(ctx, keyData, callback);
            } catch (e) {
                if (e instanceof MslEncoderException)
                    throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "keyrequestdata " + keyRequestDataMo, e);
                throw e;
            }
        });
    };
    
    // Exports.
    module.exports.parse = KeyRequestData$parse;
})(require, (typeof module !== 'undefined') ? module : mkmodule('KeyRequestData'));
